### PR TITLE
homed: also support minimizing/maximizing home dirs when creating them

### PR DESCRIPTION
**Pull request by poettering**
This hooks up the work done in #21411 + #21391 also with the "create" home dir operation, i.e. allows creating maximized/minimized home dirs right away.

It's mostly just reusing the already existing code at one more place, and is hence pretty short.

### CODE REVIEW

**Comment by poettering** [comment]
Split out of #21253

File: src/home/homework-luks.c [hunk 1]
```diff
@@ -63,6 +63,7 @@
                 _x > UINT64_MAX - 4095U ? UINT64_MAX : (_x + 4095U) & ~UINT64_C(4095); \
         })
 
+static int resize_image_loop(UserRecord *h, HomeSetup *setup, uint64_t old_image_size, uint64_t new_image_size, uint64_t *ret_image_size);
```
Comment (bluca) [inline src/home/homework-luks.c:70]: any reason to avoid moving the caller further down?

File: src/home/homework-luks.c [hunk 2]
```diff
@@ -63,6 +63,9 @@
                 _x > UINT64_MAX - 4095U ? UINT64_MAX : (_x + 4095U) & ~UINT64_C(4095); \
         })
 
+/* How much larger will the image on disk be than the fs inside it, i.e. the space we pay for the GPT and LUKS2 envelope. */
+#define GPT_LUKS2_OVERHEAD UINT64_C(18874368)
```
Comment (bluca) [inline src/home/homework-luks.c:68]: It's a very specific number, is there a rationale for it? Would be good to specify the source in the comment above

File: src/home/homework-luks.c [hunk 3]
```diff
@@ -1950,6 +1955,8 @@ static int calculate_initial_image_size(UserRecord *h, int image_fd, const char
         }
 
         lower_boundary = minimal_size_by_fs_name(fstype);
+        if (lower_boundary != UINT64_MAX)
+                lower_boundary += GPT_LUKS2_OVERHEAD;
```
Comment (bluca) [inline src/home/homework-luks.c:1961]: overflow check?

**Review by bluca — APPROVED** [review]
lgtm, just a couple minor comments

File: src/home/homework-luks.c (hunk 3 above)
Comment (poettering) [reply src/home/homework-luks.c:1961]: wel, both are constant "lower" boundaries if you so will, if they'd overflow we'd be in deep trouble... ;-)

But yeah, makes sense to encode that fact in an assert()

File: src/home/homework-luks.c (hunk 2 above)
Comment (poettering) [reply src/home/homework-luks.c:68]: this is simply what I measured on current cryptsetup, and it kinda matches docs, i.e. they say 16M JSON area plus some additional binary headers.

Comment (poettering) [reply src/home/homework-luks.c:68]: (will add comment)

File: src/home/homework-luks.c (hunk 1 above)
Comment (poettering) [reply src/home/homework-luks.c:70]: to keep this together with resize_image_fs(). i.e. the two functions are closely related: one is the bisection loop when growing, i.e. when we first to grow the backing file to the max the underlying fs allows, and the other is the bisection loop when shrinking, i.e. when we first try to shrink the inner fs to minimum it allows. I think keeping them together, along with the rest of the resize code makes sense.

**Comment by poettering** [comment]
Force pushed new version, with the suggested changes. Upgrading label.

**Comment by poettering** [comment]
neat! ubuntu amd64 CI passed. let's merge!

**Pull request merged**